eventconsumer/src/main/scala/com/gu/notifications/events/AthenaLambda.scala

```diff
@@ -0,0 +1,129 @@
+package com.gu.notifications.events
+
+import java.time.{Duration, ZonedDateTime}
+import java.util.concurrent.{CompletableFuture, CompletionStage, ConcurrentLinkedQueue}
+
+import com.gu.notifications.events.AthenaLambda.athenaAsyncClient
+import com.gu.notifications.events.dynamo.DynamoReportUpdater
+import com.gu.notifications.events.model.{AggregationCounts, EventAggregation, NotificationReportEvent, PlatformCount}
+import org.apache.logging.log4j.{LogManager, Logger}
+import software.amazon.awssdk.auth.credentials.{AwsCredentialsProviderChain, DefaultCredentialsProvider, ProfileCredentialsProvider}
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.athena.AthenaAsyncClient
+import software.amazon.awssdk.services.athena.model.{GetQueryExecutionRequest, GetQueryExecutionResponse, GetQueryResultsRequest, GetQueryResultsResponse, QueryExecutionContext, QueryExecutionState, ResultConfiguration, StartQueryExecutionRequest, StartQueryExecutionResponse}
+
+import scala.collection.JavaConverters._
+import scala.compat.java8.FutureConverters
+import scala.concurrent.ExecutionContext
+
+object AthenaLambda {
+  val reportingWindow = Duration.ofHours(3)
+  val athenaAsyncClient = AthenaAsyncClient.builder()
+    .region(Region.EU_WEST_1)
+    .credentialsProvider(AwsCredentialsProviderChain.builder()
+      .addCredentialsProvider(ProfileCredentialsProvider.builder()
+        .profileName("mobile")
+        .build())
+      .addCredentialsProvider(DefaultCredentialsProvider.builder().build())
+      .build())
+    .build()
+}
+
+case class Query(database: String, queryString: String, outputLocation: String)
+
+class AthenaLambda {
+
+  import ExecutionContext.Implicits.global
+  private val envDependencies = new EnvDependencies
+  private val logger: Logger = LogManager.getLogger(classOf[AthenaLambda])
+  private val dynamoReportUpdater = new DynamoReportUpdater(envDependencies.stage)
+
+
+  def route(query: Query, startOfReportingWindow: ZonedDateTime): CompletableFuture[Void] = {
+    startQuery(query)
+      .thenComposeAsync(fetchQueryResponse)
+      .thenComposeAsync(updateDynamoIfRecent(_, startOfReportingWindow))
+      .thenAcceptAsync((aggregationCounts: AggregationCounts) => {
+        logger.info(s"Aggregation counts $aggregationCounts")
+        if (aggregationCounts.failure > 0) {
+          throw new RuntimeException("Failures")
+        }
+      })
+  }
+
+  def startQuery(query: Query): CompletableFuture[String] = {
+    val startQueryAndGetId: CompletableFuture[String] = executeQuery(query).thenApplyAsync((startQueryExecutionResponse: StartQueryExecutionResponse) => startQueryExecutionResponse.queryExecutionId())
+    startQueryAndGetId.thenComposeAsync((executionId: String) =>
+      waitUntilQueryCompletes(executionId).thenApplyAsync((_: Void) => executionId)
+    )
+  }
+
+
+  private def waitUntilQueryCompletes(id: String): CompletableFuture[Void] = {
+    Thread.sleep(1000)
```

Comment:
Would this be of any use?
https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/

Reply:
It looks like Waiters have been relegated to the test-utils ☹️

Reply:
It might be a different one, are we using the sdk v2?
https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/waiters/package-summary.html